tests/testlineargradientbrush.c

```diff
@@ -0,0 +1,1313 @@
+{
+    GpStatus status;
+    GpLineGradient *brush;
+    GpRectF rect = { 1, 3, 1, 2 };
+    ARGB blend[2];
+    REAL positions[2];
+
+    GdipCreateLineBrushFromRect (&rect, 10, 11, LinearGradientModeHorizontal, WrapModeTile, &brush);
+    
+    status = GdipGetLinePresetBlend (brush, blend, positions, 2);
+    assert (status == GenericError);
+    
+    status = GdipGetLinePresetBlend (NULL, blend, positions, 1);
+    assert (status == InvalidParameter);
+
+    status = GdipGetLinePresetBlend (brush, NULL, positions, 1);
+    assert (status == InvalidParameter);
+
+    status = GdipGetLinePresetBlend (brush, blend, NULL, 1);
+    assert (status == InvalidParameter);
+
+    status = GdipGetLinePresetBlend (brush, blend, positions, 1);
+    assert (status == InvalidParameter);
+
+    status = GdipGetLinePresetBlend (brush, blend, positions, 0);
+    assert (status == InvalidParameter);
+
+    status = GdipGetLinePresetBlend (brush, blend, positions, -1);
+    assert (status == InvalidParameter);
+
+    GdipDeleteBrush ((GpBrush *) brush);
+}
+
+static void test_setLinePresetBlend ()
+{
+    GpStatus status;
+    GpLineGradient *brush;
+    GpRectF rect = { 1, 3, 1, 2 };
+
+    ARGB blend2[2] = { 1, 0 };
+    REAL positions2[2] = { 0, 1.0f };
+    ARGB destBlend2[2];
+    REAL destPositions2[2];
+
+    ARGB blend3[3] = { 1, 2, 3 };
+    REAL positions3[3] = { 0, 0.5f, 1.0f };
+    ARGB destBlend3[3];
+    REAL destPositions3[3];
+
+    REAL invalidPositions1[2] = { 0.5, 1 };
+    REAL invalidPositions2[2] = { 0, 0.5 };
+
+    REAL lineBlend[3] = { 1, 2, 3 };
+    REAL linePositions[3] = { 0, 2, 1 };
+    INT blendCount;
+
+    GdipCreateLineBrushFromRect (&rect, 10, 11, LinearGradientModeHorizontal, WrapModeTile, &brush);
+
+    // Count of 2.
+    status = GdipSetLinePresetBlend (brush, blend2, positions2, 2);
+    assert (status == Ok);
+
+    status = GdipGetLinePresetBlend (brush, destBlend2, destPositions2, 2);
+    assert (status == Ok);
+    assert (destBlend2[0] == 1);
+    assert (destBlend2[1] == 0);
+    assert (destPositions2[0] == 0);
+    assert (destPositions2[1] == 1);
+
+    // Count of 3.
+    status = GdipSetLinePresetBlend (brush, blend3, positions3, 3);
+    assert (status == Ok);
+
+    status = GdipGetLinePresetBlend (brush, destBlend3, destPositions3, 3);
+    assert (status == Ok);
+    assert (destBlend3[0] == 1);
+    assert (destBlend3[1] == 2);
+    assert (destBlend3[2] == 3);
+    assert (destPositions3[0] == 0);
+    assert (destPositions3[1] == 0.5);
+    assert (destPositions3[2] == 1);
+
+    // Should clear the existing blend.
+    status = GdipSetLineBlend (brush, lineBlend, linePositions, 3);
+    assert (status == Ok);
+
+    status = GdipSetLinePresetBlend (brush, destBlend2, destPositions2, 2);
+    assert (status == Ok);
+
+    status = GdipGetLineBlendCount (brush, &blendCount);
+    assert (status == Ok);
+    // FIXME: this returns 2 with GDI+.
+    // assert (blendCount == 2);
```

Comment:
I believe this is a bug in libgdiplus but am unsure - @akoeplinger what do you think!

Reply:
it sounds like a bug in libgdiplus to me as well.